### PR TITLE
cortexm_common: add FPU support for cortex-m4f and cortex-m7

### LINKED ISSUE #1366: FPU support for the Cortex-M4F family
based on the introduction of support for the Cortex-M4 familiy (#1191), support for the FPU has to be added. This means adjusting the context-switching code by including the FPU registers.

### PR DESCRIPTION
**Pull request by vincent-d**
Fixes #1366.

Not sure if every cases are covered, but it works with nucleo-f401 and nucleo-f413.
I added a test app based on `thread_msg` which uses float.

No hard fault so far and float results seem OK with different threads.

### CODE REVIEW

**Comment by haukepetersen** [comment]
Nice! Looks great. I think it might make sense to add a fpu feature, so we can do nicer filtering when building the test?!

**Comment by kaspar030** [comment]
The test runs those three threads in parallel, right?

Maybe it would be useful to crazily switch between them, e.g., using a periodic timer?

**Comment by vincent-d** [comment]
@haukepetersen I don't really know what it implies, so maybe we could. But I don't really like the Makefile stuff to disable the feature, so if you have a better idea, don't hesitate to give it.

@kaspar030 you're right, we should try to switch in the middle of the float calculations, I'll try to improve the test app

**Comment by vincent-d** [comment]
I added a patch to cmsis-dsp to include cpu_conf.h in order to fix the build issues

**Comment by vincent-d** [comment]
Test app updated to switch threads more agressively

**Comment by vincent-d** [comment]
Fixed CI issues

**Comment by kYc0o** [comment]
@vincent-d Can you squash to enable the CI?

**Comment by vincent-d** [comment]
squashed

File: cpu/Makefile.include.cortexm_common [hunk 1]
```diff
@@ -58,13 +58,20 @@ export CFLAGS += -DCPU_ARCH_$(ARCH)
...
+        export CFLAGS_FPU ?= -mfloat-abi=soft
+    else
+        # clang assumes there is an FPU
+        ifneq (llvm,$(TOOLCHAIN))
+            export CFLAGS_FPU ?= -mfloat-abi=hard -mfpu=fpv4-sp-d16
+        endif
+    endif
+    export MCPU := cortex-m4
+else
 CFLAGS_FPU ?= -mfloat-abi=soft
```
Comment (kYc0o) [inline cpu/Makefile.include.cortexm_common]: The default behaviour is then soft float?

File: tests/thread_float/Makefile [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+include ../Makefile.tests_common
+
+BOARD_INSUFFICIENT_MEMORY := airfy-beacon calliope-mini cc2650stk chronos maple-mini \
+                             mbed_lpc1768 microbit msb-430 msb-430h nrf51dongle \
+                             nrf6310 nucleo32-f031 nucleo32-f042 nucleo32-f303 \
+                             nucleo-f030 nucleo-f070 nucleo-f072 nucleo-f103 \
+                             nucleo-f334 opencm9-04 pca10000 pca10005 spark-core \
+                             stm32f0discovery weio yunjia-nrf51822
+
+DISABLE_MODULE += auto_init
```
Comment (kYc0o) [inline tests/thread_float/Makefile]: Why auto_init should be disabled?

Comment (vincent-d) [inline tests/thread_float/Makefile]: This is actually a good question :D

I copied/paste another test, so I guess we could remove that line and enable auto_init.

File: cpu/Makefile.include.cortexm_common (hunk 1 above)
Comment (vincent-d) [inline cpu/Makefile.include.cortexm_common]: No, this else is for everything **not** cortex-m4f

Comment (kYc0o) [inline cpu/Makefile.include.cortexm_common]: OK, I misread the `ifneq (llvm,$(TOOLCHAIN))`. Thanks!

File: cpu/Makefile.include.cortexm_common [hunk 3]
```diff
@@ -58,13 +58,20 @@ export CFLAGS += -DCPU_ARCH_$(ARCH)
...
-#       code is adjusted to take care of FPU registers
-#export CFLAGS_FPU += -mfloat-abi=hard -mfpu=fpv4-sp-d16
-export MCPU := cortex-m4
-endif
+    USEMODULE += cortexm_fpu
+    ifneq (,$(filter cortexm_fpu,$(DISABLE_MODULE)))
+        export CFLAGS_FPU ?= -mfloat-abi=soft
+    else
+        # clang assumes there is an FPU
+        ifneq (llvm,$(TOOLCHAIN))
```
Comment (gebart) [inline cpu/Makefile.include.cortexm_common]: It's there any reason to not give these flags to clang, regardless of the default setting? Does clang give an error if you specify hard float? Toolchains are usually quite configurable and the defaults may not always be the same.

Comment (vincent-d) [inline cpu/Makefile.include.cortexm_common]: Wihtout this ifneq, the following error is generated:
```
src/RIOT/cpu/cortexm_common/include/cmsis_gcc.h:344:54: error: unknown register name 'vfpcc' in asm
  __ASM volatile ("VMSR fpscr, %0" : : "r" (fpscr) : "vfpcc");
                                                     ^
1 error generated.
```

clang version : `3.8.1-12ubuntu1`

**Comment by vincent-d** [comment]
Auto_init no longer disabled.

**Comment by vincent-d** [comment]
- rebased
- improved `Makefile.onclude.cortexm_common` to avoid having a warning when disabling `cortexm_fpu` module

**Comment by vincent-d** [comment]
Rebased and added support for cortex-m7. Untested on cortex-m7, though.

**Comment by vincent-d** [comment]
Fixed cortex-m7 options (`-mfpu=fpv5-d16`). Tested on stm32f7discovery.

File: tests/thread_float/main.c [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    float f, init;
+
+    printf("THREAD %" PRIkernel_pid " start\n", thread_getpid());
+
+    init = 1.0 * thread_getpid();
+    f = init;
+
+    while (1) {
+        for (int i = 0; i < 100000; i++) {
```
Comment (MrKevinWeiss) [inline tests/thread_float/main.c]: Seems murdock doesn't like this too much, maybe specify uint32_t or limit it to 32767.  If you do that I will test then hopefully merge now that I am a maintainer ;)

Comment (MrKevinWeiss) [reply tests/thread_float/main.c]: I think you hit the wrong for loop, do the one on line 83.

Comment (vincent-d) [reply tests/thread_float/main.c]: Sorry, I had fixed the other one, this one should be fixed now as well

Comment (bergzand) [reply tests/thread_float/main.c]: On the AVR (among others) an `int` is only 16 bits, so this comparison is always true for those architectures.

Comment (vincent-d) [reply tests/thread_float/main.c]: @bergzand I switch to `unsigned long`

File: tests/thread_float/Makefile [hunk 5]
```diff
@@ -0,0 +1,16 @@
+APPLICATION = thread_float
+include ../Makefile.tests_common
+
+BOARD_INSUFFICIENT_MEMORY := airfy-beacon calliope-mini cc2650stk chronos maple-mini \
```
Comment (MrKevinWeiss) [inline tests/thread_float/Makefile]: Seems like murdock is still not happy, please add the following boards:
nucleo-f031k6
nucleo-f042k6

than maybe squash and everything should be good.

File: tests/thread_float/main.c [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+#define OFFSET (10 * XTIMER_BACKOFF)
+
+static mutex_t lock = MUTEX_INIT;
+
+static void timer_cb(void *arg)
+{
+    (void) arg;
+
+    xtimer_set(&timer, OFFSET);
+    sched_switch(THREAD_PRIORITY_MAIN - 2);
```
Comment (kaspar030) [inline tests/thread_float/main.c]: shouldn't "thread_yield()``` do the trick here?

Comment (MrKevinWeiss) [reply tests/thread_float/main.c]: I tested with thread_yield() and it seems to work ok!
```
2018-06-01 15:58:12,754 - INFO # THREADS CREATED
2018-06-01 15:58:12,755 - INFO # 
2018-06-01 15:58:12,756 - INFO # THTTRHHERREEAAADDD   55 5 s ststatraatrr
2018-06-01 15:58:12,757 - INFO # tt
2018-06-01 15:58:12,758 - INFO # 
2018-06-01 15:58:12,898 - INFO # T(5): 141.521576
2018-06-01 15:58:12,899 - INFO # T(3): 141.466812
2018-06-01 15:58:13,042 - INFO # T(5): 141.559769
2018-06-01 15:58:13,049 - INFO # T(3): 141.490616
2018-06-01 15:58:13,192 - INFO # T(5): 141.605148
2018-06-01 15:58:13,198 - INFO # T(3): 141.521576
2018-06-01 15:58:13,343 - INFO # T(5): 141.657516
2018-06-01 15:58:13,344 - INFO # T(3): 141.559769
2018-06-01 15:58:13,492 - INFO # T(5): 141.717255
2018-06-01 15:58:13,494 - INFO # T(3): 141.605148
```

**Comment by MrKevinWeiss** [comment]
@kaspar030 @vincent-d Should we try to push this one forward?

**Comment by kaspar030** [comment]
Yes! @vincent-d could you rebase once more?

**Comment by aabadie** [comment]
@vincent-d, just a rebase is missing. I guess you can also squash.

**Comment by MrKevinWeiss** [comment]
@vincent-d looks like arduino-uno and arduino-duemilanove needs insufficient memory to make murdock happy...

I would also like to retest quickly.

**Comment by MrKevinWeiss** [comment]
Test is the same.  @kaspar030 are the changes sufficient for you?

**Comment by vincent-d** [comment]
Fixed and amended directly

**Comment by MrKevinWeiss** [comment]
@vincent-d Thanks for the fast response. :+1: 

We wait for @kaspar030 to dismiss at this point.

**Comment by kaspar030** [comment]
I quickly checked code size, memory consumption and context switch performance. code & memory is not worth talking about, context switch performance is down 2%, which is a lot less than I expected and completely acceptable.

ACK & go.

Thanks @all and especially to @vincent-d for the work and patience!

**Comment by benpicco** [comment]
It would be great if the test had a `tests/01-run.py` to verify the correct output of the test.
I'm a bit at a loss trying to figure out what is the correct behavior to expect from the test.

**Pull request merged**